### PR TITLE
Made two specializations for affine transform easier to understand. Added AVX-512 for the specialization for small inputs

### PR DESCRIPTION
**Pull request by maximmasiutin**
Added use of AVX-512 for affine transform for the small input specialization. Comments updated. This pull request is based on the earlier pull request https://github.com/official-stockfish/Stockfish/pull/4500/ and produces identical binary. The change is in the presentation of the source code only, as suggested by @Sopel97

### CODE REVIEW

**Comment by Sopel97** [comment]
if you remove the now unused constants at the class scope then it lgtm

**Comment by Ipmanchess** [comment]
My Tournament  2023
                                     
1   Stockfish 20230407 avx512    +9  +49/=927/-24 51.25%  512.5/1000
2   Stockfish 15.1 avx512        -9  +24/=927/-49 48.75%  487.5/1000

Engine                 Depth       Time   Games     Moves  Average Forfeit  Book Depth     MIDG   EARLY    ENDG    LATE
Stockfish 15.1 avx51   31.83   17:35:55    1000     62889    1.01     0     2979  2.98    25.11 | 24.89 | 33.53 | 49.29
Stockfish 20230407 a   31.67   17:22:50    1000     62864    1.00     0     2980  2.98    24.17 | 24.19 | 34.38 | 49.68

TC: 10s+1s

**Comment by vondele** [comment]
Can you fix the CI errors / compiler warnings ?

**Comment by vondele** [comment]
If this impacts avx512 code generation, do you have a speedup benchmark that illustrates the impact?

**Comment by maximmasiutin** [comment]
vnni512 and avx512 speedup is about 0.3% on average depending on CPU, mode of use, etc. Raw measurements ("stockfish bench", "pyshtest") are available at http://www.masiutin.net/measurements-avx512-affine-transform.7z

With "stockfish bench" limited by `movetime` 10 minutes, 64 threads, hash=8192 on Xeon 8272CL (72 cores VM, hyperthreading disabled, one physical core per logical thread) NPS gain was 0.81% on ARCH=x86-64-avx512. On other CPUs (AMD Ryzen 7 7700) or other ARCHs there was no gain in "stockfish bench" results, or the results were inconclusive.

With pyshbench 500 iterations (single-threaded) gain was on Xeon 8272CL (72 cores VM, hyperthreading disabled) 1.31% on ARCH=x86-64-vnni512, and 0.24% on ARCH=x86-64-avx512, and on Core i7-1065G7 physical (notebook) 0.39% on ARCH=x86-64-vnni512. On other CPUs or ARCHs there was no gain in pyshbench results.

As for the cutechess-cli matches, Intel Xeon 8272CL (Threads=70) 3s+0.02 hash=1024 on 2000 game pairs score is 542 - 522 - 936; 2s+0.1s hash=1024 score is 558 - 544 - 898 ; 30s+1s hash=8192 200 game pairs score is 50 - 49 - 101; on AMD Ryzen 7 7700 Zen 4 (Threads=16) 30s+1s hash=1024 score is 63 - 53 - 84

Impacted functions are marked red in the screenshot https://user-images.githubusercontent.com/8037982/230185856-fc558fbb-fde4-47db-ba39-e54c7869cd56.png (propagate) which combined take around 1% after optimization in the patch, so the overall impact is not big.

**Comment by maximmasiutin** [comment]
> Can you fix the CI errors / compiler warnings ?

Which errors do you speak about? Can you please quote? I see that there we no errors. Compilation was just cancelled.

**Comment by maximmasiutin** [comment]
@vondele - if you mean errors on MacOS, let me fix that now.

**Comment by Sopel97** [comment]
```
In file included from benchmark.cpp:26:
In file included from ./position.h:32:
In file included from ./nnue/nnue_accumulator.h:24:
In file included from ./nnue/nnue_architecture.h:30:
./nnue/layers/affine_transform.h:495:34: error: unused variable 'InputSimdWidth' [-Werror,-Wunused-variable]
      static constexpr IndexType InputSimdWidth = SimdWidth;
                                 ^
1 error generated.
make[1]: *** [benchmark.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [build] Error 2
```
read my previous comment

**Comment by maximmasiutin** [comment]
Thank you! I removed this line 
```
static constexpr IndexType InputSimdWidth = SimdWidth;
```
as it is not needed. I made rebase squash.

**Comment by maximmasiutin** [comment]
All checks have passed now

**Comment by Sopel97** [comment]
```
Model name:          Intel(R) Xeon(R) Gold 6230R CPU @ 2.10GHz
CPU MHz:             4000.000
CPU max MHz:         4000.0000
CPU min MHz:         1000.0000
BogoMIPS:            4200.00
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            36608K
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr
 sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology non
stop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca s
se4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault ep
b cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid
 ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt
 clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_l
ocal dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
```

1. perf stat for avx512
```
perf stat -- ./stockfish_master_avx512 bench 16 1 16 default depth NNUE

===========================
Total time (ms) : 7063
Nodes searched  : 10230470
Nodes/second    : 1448459

 Performance counter stats for './stockfish_master_avx512 bench 16 1 16 default depth NNUE':

          7,118.95 msec task-clock                #    0.999 CPUs utilized
               113      context-switches          #   15.873 /sec
                 4      cpu-migrations            #    0.562 /sec
             6,767      page-faults               #  950.561 /sec
    24,737,430,195      cycles                    #    3.475 GHz
    32,347,555,644      instructions              #    1.31  insn per cycle
     3,121,272,628      branches                  #  438.445 M/sec
       191,346,350      branch-misses             #    6.13% of all branches

       7.123493998 seconds time elapsed

       7.062464000 seconds user
       0.042904000 seconds sys



perf stat -- ./stockfish_patch_avx512 bench 16 1 16 default depth NNUE

===========================
Total time (ms) : 7056
Nodes searched  : 10230470
Nodes/second    : 1449896

 Performance counter stats for './stockfish_patch_avx512 bench 16 1 16 default depth NNUE':

          7,112.43 msec task-clock                #    0.999 CPUs utilized
               114      context-switches          #   16.028 /sec
                 0      cpu-migrations            #    0.000 /sec
             6,803      page-faults               #  956.494 /sec
    24,608,121,009      cycles                    #    3.460 GHz
    32,079,848,858      instructions              #    1.30  insn per cycle
     3,121,284,324      branches                  #  438.849 M/sec
       191,654,817      branch-misses             #    6.14% of all branches

       7.116469438 seconds time elapsed

       7.056069000 seconds user
       0.040924000 seconds sys
```

2. perf stat for vnni512
```
perf stat -- ./stockfish_master_vnni512 bench 16 1 16 default depth NNUE

===========================
Total time (ms) : 6957
Nodes searched  : 10230470
Nodes/second    : 1470528

 Performance counter stats for './stockfish_master_vnni512 bench 16 1 16 default depth NNUE':

          7,012.77 msec task-clock                #    0.999 CPUs utilized
               119      context-switches          #   16.969 /sec
                 0      cpu-migrations            #    0.000 /sec
             6,801      page-faults               #  969.802 /sec
    24,251,212,269      cycles                    #    3.458 GHz
    30,393,215,347      instructions              #    1.25  insn per cycle
     3,108,934,642      branches                  #  443.325 M/sec
       193,159,942      branch-misses             #    6.21% of all branches

       7.017369206 seconds time elapsed

       6.958084000 seconds user
       0.042935000 seconds sys



perf stat -- ./stockfish_patch_vnni512 bench 16 1 16 default depth NNUE

===========================
Total time (ms) : 7095
Nodes searched  : 10230470
Nodes/second    : 1441926

 Performance counter stats for './stockfish_patch_vnni512 bench 16 1 16 default depth NNUE':

          7,152.36 msec task-clock                #    0.999 CPUs utilized
               115      context-switches          #   16.079 /sec
                 1      cpu-migrations            #    0.140 /sec
             6,782      page-faults               #  948.219 /sec
    24,372,482,430      cycles                    #    3.408 GHz
    30,246,201,151      instructions              #    1.24  insn per cycle
     3,108,500,541      branches                  #  434.612 M/sec
       191,492,364      branch-misses             #    6.16% of all branches

       7.156899116 seconds time elapsed

       7.098922000 seconds user
       0.038928000 seconds sys
```

3. bench for avx512
```
python3 pyshbench ./stockfish_master_avx512 ./stockfish_patch_avx512 100

Result of 100 runs
==================
base (...aster_avx512) =    1862156  +/- 9764
test (...patch_avx512) =    1857883  +/- 9372
diff                   =      -4273  +/- 4239

speedup        = -0.0023
P(speedup > 0) =  0.0243
```

4. bench for vnni512
```
python3 pyshbench ./stockfish_master_vnni512 ./stockfish_patch_vnni512 100
Result of 100 runs
==================
base (...ster_vnni512) =    1890462  +/- 8726
test (...atch_vnni512) =    1878823  +/- 8132
diff                   =     -11640  +/- 2496

speedup        = -0.0062
P(speedup > 0) =  0.0000
```


TLDR; avx512 perf less cycles, vnni512 perf more cycles (at least on this cpu). bench suggests it may be slightly worse, at least on this machine

Though, personally I don't require an improvement from this patch, just not a regression, since it does what should have been done from the start and it now allows us to attempt removing the overload for large affine layers. (Which was pretty much only necessary due to previous architecture layer sizes and was already disabled for <AVX512)

**Comment by maximmasiutin** [comment]
I've checked on AMD Ryzen 7 7700 Physical CPU (turbo up to 5.3GHz) and Intel Xeon Platinum 8370C (2.80GHz, no turbo). On AMD, patch has fewer instructions but longer time because of lower frequency comparing to master, probably due to higher energy consumption of 512-bit VNNI code comparing to that of 256-bit. On Intel Xeon Platinum 8370C, the frequency does not drop and stays at 2.80GHz, and the patch is always faster.

Raw results at C:\!!!\bench-9\2023-04-08-patch-benchmarks.7z

**CPU #1:** AMD Ryzen 7 7700 Physical; RAM 64G DDR5

flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d

bugs		: sysret_ss_attrs spectre_v1 spectre_v2 spec_store_bypass



**CPU #2:** Intel Xeon Platinum 8370C Azure VM

flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti tpr_shadow vnmi ept vpid fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap clflushopt avx512cd avx512bw avx512vl xsaveopt xsavec xsaves md_clear

vmx flags	: vnmi invvpid tsc_offset vtpr ept vpid

bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa itlb_multihit mmio_stale_data

Perf cycles, instructions, branches and branch-misses are not supported


patch: git commit `024206cc8e2eb587c335bb4dd7477930e811fda1`
master: git commit `510aca1ef62279dc35941fa45ed61fb9d3796f10`
git repo: https://github.com/maximmasiutin/Stockfish

clang: version 17, CXXFLAGS="-march=znver4 -mtune=znver4" (for AMD) and "-march=native -mtune=native" for Intel
gcc: version 12.2, CXXFLAGS="-march=native -mtune=native"



command: `perf stat -- ./stockfish bench 16 1 16 default depth NNUE`


# clang

## x86-64-avx512

### master 

#### AMD
```
===========================
Total time (ms) : 3514
Nodes searched  : 10230470
Nodes/second    : 2911346

 Performance counter stats for './stockfish-x86-64-avx512-master-clang bench 16 1 16 default depth NNUE':

          3,544.78 msec task-clock                #    1.000 CPUs utilized          
               119      context-switches          #   33.571 /sec                   
                 4      cpu-migrations            #    1.128 /sec                   
            12,684      page-faults               #    3.578 K/sec                  
    18,889,390,042      cycles                    #    5.329 GHz                      (83.11%)
        30,549,128      stalled-cycles-frontend   #    0.16% frontend cycles idle     (83.28%)
     2,160,023,194      stalled-cycles-backend    #   11.44% backend cycles idle      (83.46%)
    30,352,396,039      instructions              #    1.61  insn per cycle         
                                                  #    0.07  stalled cycles per insn  (83.42%)
     2,923,258,568      branches                  #  824.667 M/sec                    (83.43%)
       135,662,097      branch-misses             #    4.64% of all branches          (83.38%)

       3.546480271 seconds time elapsed

       3.525712000 seconds user
       0.020009000 seconds sys
```
#### Intel
```
===========================
Total time (ms) : 6639
Nodes searched  : 10230470
Nodes/second    : 1540965

 Performance counter stats for './stockfish-x86-64-avx512-patch-clang bench 16 1 16 default depth NNUE':

           6695.11 msec task-clock                #    0.999 CPUs utilized          
               114      context-switches          #   17.027 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
              6790      page-faults               #    1.014 K/sec                  

       6.699401559 seconds time elapsed

       6.656380000 seconds user
       0.039978000 seconds sys

```



### patch

#### AMD
```
===========================
Total time (ms) : 3520
Nodes searched  : 10230470
Nodes/second    : 2906383

 Performance counter stats for './stockfish-x86-64-avx512-patch-clang bench 16 1 16 default depth NNUE':

          3,551.49 msec task-clock                #    1.000 CPUs utilized          
               111      context-switches          #   31.254 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
            12,684      page-faults               #    3.571 K/sec                  
    18,922,483,025      cycles                    #    5.328 GHz                      (83.20%)
        29,641,931      stalled-cycles-frontend   #    0.16% frontend cycles idle     (83.24%)
     2,112,707,942      stalled-cycles-backend    #   11.17% backend cycles idle      (83.39%)
    30,092,935,631      instructions              #    1.59  insn per cycle         
                                                  #    0.07  stalled cycles per insn  (83.38%)
     2,915,558,842      branches                  #  820.940 M/sec                    (83.46%)
       135,706,399      branch-misses             #    4.65% of all branches          (83.41%)

       3.552938980 seconds time elapsed

       3.532272000 seconds user
       0.020001000 seconds sys
```
#### Intel
```
===========================
Total time (ms) : 6480
Nodes searched  : 10230470
Nodes/second    : 1578776

 Performance counter stats for './stockfish-x86-64-vnni512-master-clang bench 16 1 16 default depth NNUE':

           6534.20 msec task-clock                #    0.999 CPUs utilized          
               113      context-switches          #   17.294 /sec                   
                 1      cpu-migrations            #    0.153 /sec                   
              6765      page-faults               #    1.035 K/sec                  

       6.538691317 seconds time elapsed

       6.503191000 seconds user
       0.032035000 seconds sys
```



## x86-64-vnni512

### master

#### AMD

```
===========================
Total time (ms) : 3552
Nodes searched  : 10230470
Nodes/second    : 2880199

 Performance counter stats for './stockfish-x86-64-vnni512-master-clang bench 16 1 16 default depth NNUE':

          3,583.20 msec task-clock                #    1.000 CPUs utilized          
               112      context-switches          #   31.257 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
            12,684      page-faults               #    3.540 K/sec                  
    18,687,776,564      cycles                    #    5.215 GHz                      (83.22%)
        28,860,868      stalled-cycles-frontend   #    0.15% frontend cycles idle     (83.18%)
     2,508,926,131      stalled-cycles-backend    #   13.43% backend cycles idle      (83.42%)
    28,473,267,327      instructions              #    1.52  insn per cycle         
                                                  #    0.09  stalled cycles per insn  (83.39%)
     2,917,378,178      branches                  #  814.181 M/sec                    (83.41%)
       135,140,480      branch-misses             #    4.63% of all branches          (83.45%)

       3.584630403 seconds time elapsed

       3.571930000 seconds user
       0.012013000 seconds sys
```
#### Intel
```
===========================
Total time (ms) : 6480
Nodes searched  : 10230470
Nodes/second    : 1578776

 Performance counter stats for './stockfish-x86-64-vnni512-master-clang bench 16 1 16 default depth NNUE':

           6534.20 msec task-clock                #    0.999 CPUs utilized          
               113      context-switches          #   17.294 /sec                   
                 1      cpu-migrations            #    0.153 /sec                   
              6765      page-faults               #    1.035 K/sec                  
   <not supported>      cycles                                                      
   <not supported>      instructions                                                
   <not supported>      branches                                                    
   <not supported>      branch-misses                                               

       6.538691317 seconds time elapsed

       6.503191000 seconds user
       0.032035000 seconds sys

```

### patch

#### AMD

```
===========================
Total time (ms) : 3489
Nodes searched  : 10230470
Nodes/second    : 2932206

 Performance counter stats for './stockfish-x86-64-vnni512-patch-clang bench 16 1 16 default depth NNUE':

          3,518.34 msec task-clock                #    1.000 CPUs utilized          
               111      context-switches          #   31.549 /sec                   
                 1      cpu-migrations            #    0.284 /sec                   
            12,686      page-faults               #    3.606 K/sec                  
    18,754,664,250      cycles                    #    5.331 GHz                      (83.19%)
        29,805,168      stalled-cycles-frontend   #    0.16% frontend cycles idle     (83.34%)
     2,410,334,668      stalled-cycles-backend    #   12.85% backend cycles idle      (83.45%)
    28,371,108,560      instructions              #    1.51  insn per cycle         
                                                  #    0.08  stalled cycles per insn  (83.42%)
     2,912,521,976      branches                  #  827.811 M/sec                    (83.33%)
       135,125,299      branch-misses             #    4.64% of all branches          (83.36%)

       3.519957436 seconds time elapsed

       3.495126000 seconds user
       0.023966000 seconds sys
```

#### Intel

```
===========================
Total time (ms) : 6502
Nodes searched  : 10230470
Nodes/second    : 1573434

 Performance counter stats for './stockfish-x86-64-vnni512-patch-clang bench 16 1 16 default depth NNUE':

           6557.71 msec task-clock                #    0.999 CPUs utilized          
               116      context-switches          #   17.689 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
              6714      page-faults               #    1.024 K/sec                  

       6.562341501 seconds time elapsed

       6.522682000 seconds user
       0.035970000 seconds sys

```



# gcc

## x86-64-avx512

### master 

#### AMD

```
===========================
Total time (ms) : 3815
Nodes searched  : 10230470
Nodes/second    : 2681643

 Performance counter stats for './stockfish-x86-64-vnni512-master-gcc bench 16 1 16 default depth NNUE':

          3,844.94 msec task-clock                #    1.000 CPUs utilized          
               114      context-switches          #   29.649 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
            12,672      page-faults               #    3.296 K/sec                  
    20,480,281,767      cycles                    #    5.327 GHz                      (83.33%)
        33,345,370      stalled-cycles-frontend   #    0.16% frontend cycles idle     (83.38%)
     1,975,570,118      stalled-cycles-backend    #    9.65% backend cycles idle      (83.29%)
    35,005,557,048      instructions              #    1.71  insn per cycle         
                                                  #    0.06  stalled cycles per insn  (83.38%)
     3,498,160,473      branches                  #  909.810 M/sec                    (83.47%)
       145,038,718      branch-misses             #    4.15% of all branches          (83.23%)

       3.846470087 seconds time elapsed

       3.821649000 seconds user
       0.024010000 seconds sys
```
#### Intel
```
===========================
Total time (ms) : 8184
Nodes searched  : 10230470
Nodes/second    : 1250057

 Performance counter stats for './stockfish-x86-64-avx512-master-gcc bench 16 1 16 default depth NNUE':

           8234.23 msec task-clock                #    0.999 CPUs utilized          
               118      context-switches          #   14.330 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
              6683      page-faults               #  811.612 /sec                   
   <not supported>      cycles                                                      
   <not supported>      instructions                                                
   <not supported>      branches                                                    
   <not supported>      branch-misses                                               

       8.239867377 seconds time elapsed

       8.199186000 seconds user
       0.036013000 seconds sys

```

### patch


#### AMD
```
===========================
Total time (ms) : 3988
Nodes searched  : 10230470
Nodes/second    : 2565313

 Performance counter stats for './stockfish-x86-64-avx512-patch-gcc bench 16 1 16 default depth NNUE':

          4,017.27 msec task-clock                #    1.000 CPUs utilized          
               113      context-switches          #   28.129 /sec                   
                 1      cpu-migrations            #    0.249 /sec                   
            12,675      page-faults               #    3.155 K/sec                  
    20,677,304,750      cycles                    #    5.147 GHz                      (83.25%)
        31,087,047      stalled-cycles-frontend   #    0.15% frontend cycles idle     (83.34%)
     1,961,492,907      stalled-cycles-backend    #    9.49% backend cycles idle      (83.42%)
    36,791,711,827      instructions              #    1.78  insn per cycle         
                                                  #    0.05  stalled cycles per insn  (83.30%)
     3,510,158,667      branches                  #  873.768 M/sec                    (83.39%)
       143,972,672      branch-misses             #    4.10% of all branches          (83.36%)

       4.019018068 seconds time elapsed

       4.002012000 seconds user
       0.016024000 seconds sys

```
#### Intel
```
===========================
Total time (ms) : 8080
Nodes searched  : 10230470
Nodes/second    : 1266147

 Performance counter stats for './stockfish-x86-64-avx512-patch-gcc bench 16 1 16 default depth NNUE':

           8133.42 msec task-clock                #    0.999 CPUs utilized          
               117      context-switches          #   14.385 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
              7297      page-faults               #  897.162 /sec                   

       8.138785937 seconds time elapsed

       8.102341000 seconds user
       0.031993000 seconds sys

```


## x86-64-vnni512

### master 

#### AMD

```
===========================
Total time (ms) : 3815
Nodes searched  : 10230470
Nodes/second    : 2681643

 Performance counter stats for './stockfish-x86-64-vnni512-master-gcc bench 16 1 16 default depth NNUE':

          3,844.94 msec task-clock                #    1.000 CPUs utilized          
               114      context-switches          #   29.649 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
            12,672      page-faults               #    3.296 K/sec                  
    20,480,281,767      cycles                    #    5.327 GHz                      (83.33%)
        33,345,370      stalled-cycles-frontend   #    0.16% frontend cycles idle     (83.38%)
     1,975,570,118      stalled-cycles-backend    #    9.65% backend cycles idle      (83.29%)
    35,005,557,048      instructions              #    1.71  insn per cycle         
                                                  #    0.06  stalled cycles per insn  (83.38%)
     3,498,160,473      branches                  #  909.810 M/sec                    (83.47%)
       145,038,718      branch-misses             #    4.15% of all branches          (83.23%)

       3.846470087 seconds time elapsed

       3.821649000 seconds user
       0.024010000 seconds sys
```
#### Intel

```
===========================
Total time (ms) : 7759
Nodes searched  : 10230470
Nodes/second    : 1318529

 Performance counter stats for './stockfish-x86-64-vnni512-master-gcc bench 16 1 16 default depth NNUE':

           7809.96 msec task-clock                #    0.999 CPUs utilized          
               119      context-switches          #   15.237 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
              6433      page-faults               #  823.691 /sec                   
   <not supported>      cycles                                                      
   <not supported>      instructions                                                
   <not supported>      branches                                                    
   <not supported>      branch-misses                                               

       7.816506987 seconds time elapsed

       7.778918000 seconds user
       0.031995000 seconds sys
```
### Patch
#### AMD

```
===========================
Total time (ms) : 4011
Nodes searched  : 10230470
Nodes/second    : 2550603

 Performance counter stats for './stockfish-x86-64-vnni512-patch-gcc bench 16 1 16 default depth NNUE':

          4,040.41 msec task-clock                #    1.000 CPUs utilized          
               115      context-switches          #   28.462 /sec                   
                 2      cpu-migrations            #    0.495 /sec                   
            12,673      page-faults               #    3.137 K/sec                  
    20,629,041,404      cycles                    #    5.106 GHz                      (83.26%)
        38,905,716      stalled-cycles-frontend   #    0.19% frontend cycles idle     (83.38%)
     1,893,397,788      stalled-cycles-backend    #    9.18% backend cycles idle      (83.41%)
    34,978,852,971      instructions              #    1.70  insn per cycle         
                                                  #    0.05  stalled cycles per insn  (83.39%)
     3,501,381,267      branches                  #  866.590 M/sec                    (83.38%)
       143,809,839      branch-misses             #    4.11% of all branches          (83.25%)

       4.041801143 seconds time elapsed

       4.017101000 seconds user
       0.024030000 seconds sys

```
#### Intel
```
===========================
Total time (ms) : 7690
Nodes searched  : 10230470
Nodes/second    : 1330360

 Performance counter stats for './stockfish-x86-64-vnni512-patch-gcc bench 16 1 16 default depth NNUE':

           7739.88 msec task-clock                #    0.999 CPUs utilized          
               120      context-switches          #   15.504 /sec                   
                 1      cpu-migrations            #    0.129 /sec                   
              7206      page-faults               #  931.022 /sec                   

       7.746228369 seconds time elapsed

       7.704981000 seconds user
       0.035985000 seconds sys
```
